### ISSUE TITLE
Tracking issue for RFC 2070: stable mechanism to specify the behavior of panic! in no-std applications 

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#44489]
### Update(2018-08-28)

- [Stabilization summary](https://github.com/rust-lang/rust/issues/44489#issuecomment-416645946).
- [FCP checkboxes](https://github.com/rust-lang/rust/issues/44489#issuecomment-398965881)

---

This is a tracking issue for the RFC "stable mechanism to specify the behavior of panic! in no-std applications " (rust-lang/rfcs#2070).

**Steps:**

- [x] Implement the RFC
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
  - [ ] The doc-comment of libcore mentions `rust_begin_panic`
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

## `fmt::Display`

Should the `Display` of `PanicInfo` format the panic information as `"panicked
at 'reason', src/main.rs:27:4"`, as `"'reason', src/main.rs:27:4"`, or simply as
`"reason"`.

## Unwinding in no-std

Is this design compatible, or can it be extended to work, with unwinding
implementations for no-std environments?

**Issues to solve before stabilization**

- [x] panic_implementation: Error message still talks about the panic_fmt lang item when missing #51341
- [x] panic_implementation: no_mangle attribute needed?  #51342 
- [x] `rust_begin_unwind` not emitted when crate-type = bin #51671

**Comment by whitequark** [comment]
> Unwinding in no-std
> 
> Is this design compatible, or can it be extended to work, with unwinding
> implementations for no-std environments?

As someone with a working unwinding implementation in a no_std environment (not with a Rust personality function though), I can answer this if someone summarizes the feature for me. I kind of lost track and I don't have time to get into every minor detail at the moment.

**Comment by yodaldevoid** [comment]
> As someone with a working unwinding implementation in a no_std environment (not with a Rust personality function though)

Sorry to bring this off topic so soon, but could you explain this? I was under the impression that unless you wanted to implement unwinding in an ad-hoc way (e.g. doing unwinding during panic by explicitly calling some function) you had to implement the `eh_personality` function.

**Comment by whitequark** [comment]
> I was under the impression that unless you wanted to implement unwinding in an ad-hoc way (e.g. doing unwinding during panic by explicitly calling some function) you had to implement the eh_personality function.

I use unwinding in a language runtime. The language has exceptions (and so unwinding), however Rust frames are never unwound, and so there's no need for the personality function.

**Comment by Zoxc** [comment]
I would like the `Location` interface to both support the ability to pass backtraces (effectively a `fn caller(&self) -> Option<Location>`) and decompression of location information which requires accessing this information in a closure passed to a `Location` function to avoid heap allocation.

**Comment by Zoxc** [comment]
We should also support no location information at all, and abstract location information (for example just an identifying number which can be used to find the relevant information in some external source).

**Comment by Zoxc** [comment]
I did not notice that `PanicInfo` and `Location` were already stable. I guess we can just deprecate `PanicInfo::location` and make it return `None` if we want some more expressive API in the future.

**Comment by Zoxc** [comment]
I'm really not a fan of using the same API for payloads and messages. I'd rather `panic!` be pure message and provide some other API for payloads. So basically I want `PanicInfo` without `location` and `payload`. I have not seen any motivation for payloads for no_std use cases either.

**Comment by Zoxc** [comment]
I kind of prefer a `PanicInfo` type with no fields or methods and just an `Display` implementation.

One problem with mixing the payload and the message is that we can't strip panic messages from the binary without changing the meaning of programs relying on payloads.

**Comment by SimonSapin** [comment]
#47687 does part of the implementation. I think after that what’s needed is:

* Change the signature of the `panic_fmt` lang item from `(msg: fmt::Arguments, file: &'static str, line: u32, col: u32) -> !` to `(&PanicInfo) -> !`. This a breaking change to an unstable feature. (Maybe also rename it to `begin_panic` or something without `fmt` in the name?)
  - When creating a `PanicInfo` in  `src/libcore/panicking.rs` to call the modified lang item, use the `&'static str` message as the payload in `panic()` and a value of a new private `struct NoPayload;` type in `panic_fmt()`
  - Make that `core::panicking::panic` function generic to accept any `P: Any + Send` payload instead of just `&'static str`.
* Add an unstable `#[panic_implementation]` proc-macro attribute gated on `#![feature(panic_implementation)]` that checks that it is on a `extern fn` with the appropriate signature, and expands to `#[lang = "panic_fmt"]`. This expansion uses stability hygiene such that the containing crate does not need the `lang_items` feature gate, only the `panic_implementation` one (until that one is stabilized)
* Change the ``error[E0152]: duplicate lang item found: `panic_fmt` `` error to have its own error number and a message that talks about duplicate `#[panic_implementation]` instead. Or maybe write new code to emit that new error when appropriate. See what the implementation of `#[global_allocator]` does.

Note however that `src/libcore/panicking.rs` contains some comments about doing a things a certain way so that so that the linker can eliminate some of the string formatting machinery to limit code size bloat when formatting is not otherwise used. It’s possible that these changes would defeat those optimizations. I don’t know what we can do to try and preserve them while implementing RFC 2070.

CC @japaric

**Comment by SimonSapin** [comment]
@Zoxc We can still propose changing RFC 2070, but a `Box<Any + Send + 'static>` payload for panics is already part of some stable APIs:

* In `std::thread::Result`, the return type of `std::thread::JoinHandle` and `std::panic::catch_unwind`
* As a parameter of `std::panic::resume_unwind`
* `PanicInfo::payload` and `std::panic::set_hook` are already stable

**Comment by Zoxc** [comment]
@SimonSapin We could just pass a `NoPayload` struct for panics without payloads though. Anyway we don't need to repeat mistakes in libcore.

**Comment by nagisa** [comment]
From what I can tell, the RFC still does not address removal of unnecessary panic information directly, however it is an extremely easy thing to do that does not require much extra design past what was already accepted.

Consider for example this scenario:

```rust
#[panic_implementation]
fn my_panic_impl(pi: &PanicInfo) -> ! {
    abort()
}
```

This is a case, where it is entirely within the power of compiler to not keep any strings, line numbers, etc. pertaining to panic locations. However as soon as the panic implementation uses *any* information:

```rust
#[panic_implementation]
fn my_panic_impl(pi: &PanicInfo) -> ! {
    abort_with_line(pi.location(),map(|l| l.line()));
}
````

the compiler is forced to keep *all* the panic information, despite the fact that only the line numbers are of any interest for that particular implementation of panicking.

---

I want to propose a following change (benefits of which are laid out at the end of this comment):

```rust
struct AnyConcreteTypeMayGoHere { line: Option<u64> }

impl ::core::panic::PanicInfoCtr for AnyConcreteTypeMayGoHere {
    const fn new_v1(message: Option<&fmt::Arguments>, location: Option<Location>, ...etc) -> Self {
        AnyConcreteTypeMayGoHere { line: location.map(|l| l.line()) }
    }
}

#[panic_implementation]
fn my_panic_impl(pi: &AnyConcreteTypeMayGoHere) -> ! { abort_with_line(pi.line) }
```

and at the location of panic rust would generate code similar to this (with `_` filled in with the type that `my_panic_impl` takes as an argument):

```rust
static PANIC_INFO: _ = <_ as ::core::panic::PanicInfoCtr>::new_v1(args, loc, ...);
my_panic_impl(&PANIC_INFO);
```

Benefits of this approach are:

1. It allows users to only keep the important bits about the panic, without relying on the compiler optimisations. `AnyConcreteTypeMayGoHere` will only retain line info even before optimisations come into play. It would also allow omitting the panic information entirely (important not only for embedded and wasm, but also closed-source code that doesn't want to expose internal information¹)
2. It is fairly extensible in backward-compatible manner without breaking the users. Consider for example addition of column numbers, which bloated various applications where code size is very important. With `AnyConcreteTypeMayGoHere` not retaining the column number in the first place, no extra bloat would’ve been introduced. Furthermore, if we wanted to add another argument to the `PanicInfo` constructor, `new_v2(...)` with default implementation calling `new_v1` would be a backward-compatible change, etc etc.

`libpanic_unwind`/`libpanic_abort` would then provide their own `#[panic_implementation]` that takes `&std::panic::PanicInfo` as an argument, and libcore would provide a few default common-sense implementations of `PanicInfoCtr` trait -- one that mimics the `std::panic::PanicInfo` as closely as feasible, one that retains no information whatsoever, etc, so that `#[panic_implementation]` could be made stable before `const fn` is. And people whom none of defaults work out can implement their own `PanicInfo` using the `const_fn` unstable feature.

¹: there were a few issues about this, but I can’t find them anymore.

**Comment by nagisa** [comment]
Status update: I realised above cannot work without being very unwieldy. Namely, the proposal above would require the `panic_implementation` to be available to all the leaf crates, before code for them could be generated. This, unlike e.g. custom allocator providers, wouldn’t be something that would work simply by linking in some stuff from anywhere in the crate graph.

**Comment by glandium** [comment]
There is an issue that IMHO, a stable mechanism for panic! in no-std applications should take into account: for at least two reasons, you can only set the panic handler in leaf crates.

One of them is issue #48661, where the linker simply removes the panic handler if it's in its own crate.

The second of them is that the `cargo test` harness pulls `std`, and the build fails with:
```
error: duplicate lang item in crate `std`: `panic_fmt`.
  |
  = note: first defined in crate `panic`.
```
and there is no way for such a crate to be built differently for `test`.

**Comment by japaric** [comment]
@glandium 

> The second of them is that the cargo test harness pulls std, and the build fails with:

This may be a problem short term but custom test framework will become available this year and with those it will be possible to write test runners than don't rely on unwinding (e.g. unit tests return Err to indicate failure) which will make it possible to test targets (e.g. thumbv7m-none-eabi) that don't have a panic implementation that unwinds.

@nagisa 

> Namely, the proposal above would require the panic_implementation to be available to all the leaf crates, before code for them could be generated.

This is probably a bad idea but would it be possible to store all panic!s in MIR format and defer the actual generation of the static variable (and the llvm-ir) until the user defined PanicInfo struct is known?

**Comment by nagisa** [comment]
Yeah, without overthinking the idea too much, I think that would be possible, however it has a hard prerequisite on MIR-only (r)libs.

**Comment by glandium** [comment]
@japaric: there are other reasons than unwinding to want to use std for tests on otherwise no_std crates. If your panic function is in a crate that contains nothing else, maybe a `#[cfg(not(test))]` on the `extern crate` can save you, but if your panic function is in a crate along with other things, there's no way out.

**Comment by nagisa** [comment]
A summary from the All Hands /wrt panic size problem.

We did discuss various approaches to handling the issue of panic size, and figured out that we’d love to have something that works in all scenarios such as non-lto builds and even debug builds. We didn’t arrive at any particular design, but we managed to reject a number of designs that wouldn’t work and also realised that the only real way to properly implement any design would rely on MIR-only rlibs.

With that in mind, this design seems to be as good as any other and, as long as we stabilise `#[panic_implementation]` carefully (i.e. the signature and behaviour, but not the exact expansion), we should be in a good place to implement a solution for panic sizes as well.

---

I’m willing to mentor implementation of the RFC. The instructions written down by @SimonSapin are part-way right; I’ll write up some new and more detailed instructions over the week or two.

**Comment by csssuf** [comment]
I'd be interested in working on this, if nobody's already on it. I'll take a look at the partial instructions now and keep an eye out for the updated ones.

**Comment by nagisa** [comment]
Sorry for the long delay!

So to implement the RFC as it is proposed (past what’s already implemented in https://github.com/rust-lang/rust/pull/47687) the following need to be done:

* Add a new language items named `panic_implementation` and `panic_info`. Remove the `panic_fmt` language item. [In this file](https://github.com/rust-lang/rust/blob/fb730d75d4c1c05c90419841758300b6fbf01250/src/librustc/middle/lang_items.rs#L297). The signature of this new `panic_implementation` language item should be as such: `fn panic_impl(&PanicInfo) -> !;`.

    I still haven’t gotten around to implementing type checking for language items, so we’ll have to implement checking for the signature the same way as it is implemented for termination trait [here](https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/mod.rs#L1130) (eww). Checks to be done here are:
    
    * Function does not terminate (its return is specified to be `-> !`;
    * Function has one argument only;
    * Function argument type is `&PanicInfo`, where `PanicInfo` is the `panic_info` language item.

* Annotate the `PanicInfo` with `panic_info` language item.

* Do not change the signature of [`libcore::panicking::panic`](https://github.com/rust-lang/rust/blob/master/src/libcore/panicking.rs#L43), and instead add another function in the same file:

    ```
    extern {
         #[lang = "panic_implementation"]
         fn panic_impl(pi: &PanicInfo) -> !;
    }
    
    /// Constructs PanicInfo and calls the `panic_implementation`.
    fn panic_payload<M: Any + Send>(msg: M, file_line_col: &(&'static str, u32, u32)) -> ! {
        panic_impl(&...)
    }
    ```

    It is worth doing things this way, because the `panic` language item is something that is relied on by the compiler internals, and it would be a pain to adjust all those calls to call a changed language item.
    
    Then, adjust the `macro_rules! panic` 2nd matcher in [`src/libcore/macros.rs`](https://github.com/rust-lang/rust/blob/fb730d75d4c1c05c90419841758300b6fbf01250/src/libcore/macros.rs#L19) to call the new `panic_payload` function.
    
    Additionally, adjust `panic` and `panic_fmt` functions in a similar manner to call the
    `panic_implementation` language item as well. Both of these functions should construct PanicInfo
    with `&NoPayload` unit struct as the payload.

* Implement the `#[panic_implementation]` attribute. One of the approaches to the implementation
   would be to use proc-macros, but AFAICT this is a pretty big hammer to use for such a tiny
   staple. The expansion should end up being trivial, after all:

    ```rust
    #[panic_implementation]
    fn my_panic_implementation(pi: &PanicInfo) -> ! {
        // body
    }
    
    // expands to
    
    #[lang="panic_implementation"]
    fn my_panic_implementation(pi: &PanicInfo) -> ! {
        // body
    }
    ```

    so all that needs to be done to make this work is to adjust [this visitor](https://github.com/rust-lang/rust/blob/fb730d75d4c1c05c90419841758300b6fbf01250/src/librustc/middle/lang_items.rs#L106) to also consider the `panic_implementation` attribute in addition to the `lang` attribute.

    This also makes reporting an error about duplicate `panic_implementations` as well as stability checking fairly easy. Adding a feature should happen [here](https://github.com/rust-lang/rust/blob/fb730d75d4c1c05c90419841758300b6fbf01250/src/libsyntax/feature_gate.rs).

* Replace the implementation of `lang = "panic_fmt"` in libstd with the newly minted
  `#[panic_implementation]`.

If you hit a roadblock or have any questions, you can find me (nagisa) and many more individuals who’d be willing to answer any questions on the #rustc IRC channel (irc.mozilla.org server).

**Comment by japaric** [comment]
PR #50338 implements `#[panic_implementation]` as per @nagisa instructions.

**Comment by japaric** [comment]
> PR #50338 implements #[panic_implementation] as per @nagisa instructions.

PR landed yesterday and the feature is available in the latest nightly.

I have added issues #51341 and #51342 to this ticket description.

**Comment by pepyakin** [comment]
There is a regression which seems related to the `panic_impl`: https://github.com/rust-lang/rust/issues/51365

**Comment by SimonSapin** [comment]
The doc-comment of the `core` crate mentions a `rust_begin_panic` and probably needs to be updated: https://doc.rust-lang.org/nightly/core/#how-to-use-the-core-library

(What about `rust_eh_personality`, is that still relevant?)

**Comment by nagisa** [comment]
@rfcbot fcp merge

This feature is one of the last few that we need(ed) to have `#[no_std]`/embedded be *really* feasible on stable versions of the compiler.

There are still implementation issues to solve before this can really be stabilised, however since these issues are "just" implementation bugs rather than inherent design flaws, they do not preclude running an FCP early.

Note, that for this feature we have punted on figuring out the issue of compiler retaining too much information about panics when a trivial `#[panic_implementation]` is used – implementing this essentially requires some sort of link-time codegen, which we aren’t getting before at least MIR-only-RLIBs.

**Comment by nagisa** [comment]
> What about rust_eh_personality, is that still relevant?

Yes, `rust_eh_personality` is important when you are trying to implement `-Cpanic=unwind`. `-Cpanic=abort` is the only thing that one can implement without extra unstable features.

**Comment by petrochenkov** [comment]
@rfcbot concern panic_thin

There's a competing RFC in flight - https://github.com/rust-lang/rfcs/pull/2305.
Supposedly there should be an explanation why the stabilized version is preferable and then https://github.com/rust-lang/rfcs/pull/2305 should be closed.

**Comment by SimonSapin** [comment]
The stated goal of that competing RFC is to avoid the code size cost of using `core::fmt`. However I think this is already possible with the current design:

```rust
#[panic_implementation]
fn panic(info: &core::panic::PanicInfo) -> ! {
    if let Some(s) = info.payload().downcast_ref::<&'static str>() {
        print(s)
    }
    loop {}
}

fn print(_s: &str) {
    // …
}
```

Creating a `PanicInfo` involves in many cases calling `core::fmt::Arguments::new_v1`, but that’s a trivial constructor with `#[inline]`. As far as I understand, the `fmt` machinery that takes a lot of code size would not be included as long as `impl Dislay for PanicInfo` is not used.

At the moment I think that this downcast to `&str` would return something with `std::panic!("foo")` but not `core::panic!("foo")`, that should be easy to fix.

**Comment by japaric** [comment]
re: Issues to solve before stabilization

#51342 and #51671 have been fixed. #51341has a fix in #51921.

**Comment by Ericson2314** [comment]
https://github.com/rust-lang/rfcs/pull/2492 an alternative of using a trait and ZST, rather than annotation, for this. Should be orthogonal to https://github.com/rust-lang/rfcs/pull/2070 vs https://github.com/rust-lang/rfcs/pull/2305 I think.

**Comment by gnzlbg** [comment]
Solving https://github.com/rust-lang/rust/issues/51981 might require changing the signature of 

```rust
fn panic_payload<M: Any + Send>(msg: M, files_lines_cols: &[(&'static str, u32, u32)]) -> ! {
    panic_impl(&...)
}
```

to take a slice.

**Comment by phil-opp** [comment]
Just as a heads up: The ["panic_implementation: no_mangle attribute needed?" issue](https://github.com/rust-lang/rust/issues/51342) still persists. Seems like the fix wasn't successful. I mentioned it in the issue but got no reaction, so I thought I mention it here too because it's wrongly marked as solved in the issue description.

**Comment by Mark-Simulacrum** [comment]
The embedded working group would like this feature to be stabilized in time for the edition, so this is a ping to remind the folks who have not yet signed off on [the FCP](https://github.com/rust-lang/rust/issues/44489#issuecomment-398965881). It does look like there's at least some unresolved concerns as well, it would be good for someone knowledgeable in this area to champion this feature.

cc @Zoxc @cramertj @michaelwoerister @nikomatsakis @petrochenkov @pnkfelix @sfackler @varkor @withoutboats @japaric

**Comment by japaric** [comment]
@Mark-Simulacrum 

> The embedded working group would like this feature to be stabilized in time for the edition

Correction: We **need** this feature to be stabilized and part of the 2018 edition. Without it none of our examples / applications would compile on 1.31.

> It does look like there's at least some unresolved concerns as well

@SimonSapin has addressed the panic_thin concern in https://github.com/rust-lang/rust/issues/44489#issuecomment-401418386.

---

@Centril has pointed out rust-lang/rfcs#2492 as an alternative to `#![panic_implementation]`. These are my thoughts on RFC 2492:

Unless there's 100% certainty that (a) we actually want to accept that RFC and that (b) we can implement it, test it and stabilize it by the 1.31-beta cutoff (2018-10-25) then I don't see it as a feasible *alternative* to `#[panic_implementation]`. Rejecting `#[panic_implementation]` in favor of RFC 2492 risks leaving support for embedded devices out of the edition release.

Even if we do accept RFC 2492 it's already too late for it to be *only* solution to the "needs / provides" problem. `#[global_allocator]`, which RFC 2492 [subsumes](https://github.com/rust-lang/rust/issues/27389#issuecomment-407519617), is already in beta and will be in stable next week so both features will have to coexist and `#[global_allocator]` will have to be deprecated in favor of RFC 2492 when it becomes stable.

In conclusion: I don't see a decision on RFC 2492 as a blocker for stabilizing `#[panic_implementation]`. If we do accept RFC 2492, `#[panic_implementation]` will be in the same boat as `#[global_allocator]` and will eventually be deprecated.

I want to end my comment with this quote from the [2018 Roadmap](https://github.com/rust-lang/rfcs/blob/master/text/2314-roadmap-2018.md#shipping):

> Shipping requires the discipline to say “no” to exciting new ideas, and to say “yes” to delivering on ones we fear are flawed. We have to iterate toward perfection, providing useful snapshots along the way.

**Comment by Ericson2314** [comment]
That all makes sense to me. I don't think my RFC has loose design ends, but implementing and stabilizing *anything* at all in half a year I think is faster than the normal, right? It's nice with editions we have a clear route to depreciating language features. Ideally I would have written my RFC a year ago right after the explicit existential types RFC (can't even say why didn't) but oh well.

**Comment by Centril** [comment]
Makes sense to me as well; Nominating for next lang team meeting to make sure everyone there agree as well. :)

**Comment by SimonSapin** [comment]
> @SimonSapin has addressed the panic_thin concern in https://github.com/rust-lang/rust/issues/44489#issuecomment-401418386.

Well, maybe. I haven’t actually checked that the `core::fmt` machinery’s code is indeed eliminated with such code. See for example https://github.com/rustwasm/team/issues/19#issuecomment-358814493

Beyond code size in terms of function bodies, I don’t know if the current API allows eliminating strings (either in the panic message/payload or in the location) when they end up not being used: https://github.com/rust-lang-nursery/embedded-wg/issues/41

**Comment by japaric** [comment]
`#[panic_implementation]` can be seen as a no_std, static version of [`std::panic::set_hook`](https://doc.rust-lang.org/std/panic/fn.set_hook.html) (stable API).

`#[alloc_error_handler]` (#51540) can be seen as a no_std, static version of [`std::alloc::set_alloc_error_hook`](https://doc.rust-lang.org/nightly/std/alloc/fn.set_alloc_error_hook.html) (unstable API).

Would it be more consistent to rename `#[panic_implementation]` to `#[panic_handler]`?

Alternatively, the attributes could be renamed to `#[*_hook]` to match the name used in the `std` API, but the word [hook](https://en.wikipedia.org/wiki/Hooking), in programming, carries the meaning of something that alters / overrides the default behavior of a program so it doesn't seem appropriate.

cc @Centril @SimonSapin

**Comment by Centril** [comment]
@japaric  I like `#[panic_handler]`, it seems consistent, short and sweet. :)

**Comment by SimonSapin** [comment]
I would prefer to specifically *not* use the same term as in `std` API because they are different. The attributes (handlers?) are mandatory (in `no_std` apps using related functionality), set at compile-time, and statically dispatched. The hooks have a default, cannot be set until run-time, can be changed during run-time, and are dynamically dispatched

**Comment by SimonSapin** [comment]
> solving the problem of payloads and panic information being kept in the binary requires pure MIR rlibs.

My reading of https://github.com/rust-lang/rust/issues/44489#issuecomment-378058031 is that MIR-only rlibs are required to eliminate unused data "in all scenarios such as non-lto builds and even debug builds". Wouldn’t (full) LTO achieve that result today?

> the #[panic_implementation] API doesn't prevent us from fixing the size problem in the future.

I think some comments suggested that separate arguments would optimize better. If any field of `PanicInfo` is used through `&PanicInfo`, none of them can be optimized out.

**Comment by SimonSapin** [comment]
It was just pointed out to me that the attribute, since it is implemented as compiler code, can look at the function’s signature and eventually accept multiple signatures (possibly generating a wrapper function if needed). So we’re not necessarily stuck `&PanicInfo` forever, which solves my concern.

I’ve formally proposed FCP to postpone https://github.com/rust-lang/rfcs/pull/2305. @petrochenkov, I think this solves your concern.

I’ll file another one for https://github.com/rust-lang/rust/issues/44489#issuecomment-409392812 but it should be easy: let’s pick one of `#[panic_implementation]` or `#[panic_handler]`:

@rfcbot concern attribute name

**Comment by mark-i-m** [comment]
I will car my vote for panic_handler

**Comment by petrochenkov** [comment]
@rfcbot resolve panic_thin

**Comment by Centril** [comment]
The lang team briefly discussed this and I think the general consensus was that the team does not want to block this on rust-lang/rfcs#2492.

**Comment by Centril** [comment]
@rfcbot poll libs compiler Can we find consensus around the name `#[panic_handler]`?

**Comment by SimonSapin** [comment]
I’ve added the attribute name to the agenda for tomorrow’s libs triage meeting.

**Comment by SimonSapin** [comment]
The libs team discussed this during today’s triage meeting. We’re happy with renaming to `panic_handler`, which is consistent with `alloc_error_handler`.

@rfcbot resolve attribute name

Anyone feels up to making a PR to rename? If that’s easy enough to implement, let’s also keep the old name for a while and make it emit a deprecation warning.

**Comment by SimonSapin** [comment]
Ping @Zoxc @michaelwoerister @nikomatsakis @pnkfelix @sfackler @withoutboats for your checkboxes in https://github.com/rust-lang/rust/issues/44489#issuecomment-398965881

**Comment by japaric** [comment]
> Anyone feels up to making a PR to rename?

I'll make a PR.

**Comment by japaric** [comment]
> I'll make a PR.

See #53619

**Comment by nikomatsakis** [comment]
@rfcbot fcp reviewed

I'm going to mark this as reviewed because it's been a while and I missed the initial FCP request (sorry about that), but for the record I'd generally prefer to see:

- a summary of the semantics to be stabilized
    - in particular any cases where things diverged from the RFC
        - it sounds like at minimum there has been some renaming
- links to test cases showing that we are testing those semantics
- a listing of bugs that remain to be fixed etc

I believe we also would like a commitment from the @rust-lang/docs team that there is enough information for them to be **able** to write docs on this feature (hopefully the RFC suffices here).

(We should formalize this process, though.)

**Comment by japaric** [comment]
As per Niko's request this is a summary of the above thread and covers what's
being stabilized, how it's tested, deviations from the RFC and remaining known
bugs:

RFC text: https://github.com/rust-lang/rfcs/blob/master/text/2070-panic-implementation.md

*What's being stabilized:*

- The `panic_handler` attribute. This attribute lets you define the behavior of
  `core::panic!`. This attribute is basically the old `panic_fmt` lang item but
  with a stable API.

*What's not being stabilized at this point:*

- Support for payloads in `core::panic!` (e.g. `panic!(42)`). This has not been
  implemented but it's backward compatible to add. Rationale was to speed up
  stabilization by not adding new functionality to `core::panic!`.

*We test:*

- That the `panic_handler` has the signature specified in the RFC:
  `fn(&PanicInfo) -> !`. Tests:
  - [`fn(PanicInfo) -> ()`][a]
  - [`fn(&'static PanicInfo) -> !`][b]
  - [`fn() -> !`][c]
  - [`fn<T>(&PanicInfo) -> !`][d]

[a]: https://github.com/rust-lang/rust/blob/ec4a752202e56a9b2d978305cb2603208d4a3ab4/src/test/ui/panic-handler/panic-handler-bad-signature-1.rs
[b]: https://github.com/rust-lang/rust/blob/ec4a752202e56a9b2d978305cb2603208d4a3ab4/src/test/ui/panic-handler/panic-handler-bad-signature-2.rs
[c]: https://github.com/rust-lang/rust/blob/ec4a752202e56a9b2d978305cb2603208d4a3ab4/src/test/ui/panic-handler/panic-handler-bad-signature-3.rs
[d]: https://github.com/rust-lang/rust/blob/ec4a752202e56a9b2d978305cb2603208d4a3ab4/src/test/ui/panic-handler/panic-handler-bad-signature-4.rs

- That only a single instance of `panic_handler` is allowed to exist in the
  crate dependency graph. Tests:
  - [`no_std` context](https://github.com/rust-lang/rust/blob/ec4a752202e56a9b2d978305cb2603208d4a3ab4/src/test/compile-fail/panic-handler-twice.rs)
  - [collision with the `panic_impl` lang item (which is what `panic_handler`
    desugars to)](https://github.com/rust-lang/rust/blob/ec4a752202e56a9b2d978305cb2603208d4a3ab4/src/test/ui/panic-handler/panic-handler-duplicate.rs)
  - [collision with the definition in `std`](https://github.com/rust-lang/rust/blob/ec4a752202e56a9b2d978305cb2603208d4a3ab4/src/test/ui/panic-handler/panic-handler-std.rs)

- That the compiler rejects binary crates that do not define `panic_handler`.
  [Test]

[Test]: https://github.com/rust-lang/rust/blob/ec4a752202e56a9b2d978305cb2603208d4a3ab4/src/test/compile-fail/panic-handler-missing.rs

- That the generated symbols have the right visibility. These are regressions
  tests for bugs that we encountered in practice. Tests:
  - [WASM exports](https://github.com/rust-lang/rust/tree/ec4a752202e56a9b2d978305cb2603208d4a3ab4/src/test/run-make/wasm-symbols-not-exported)
  - [WASM imports](https://github.com/rust-lang/rust/tree/ec4a752202e56a9b2d978305cb2603208d4a3ab4/src/test/run-make/wasm-symbols-not-imported)
  - [`panic_handler` in top binary crate](https://github.com/rust-lang/rust/tree/ec4a752202e56a9b2d978305cb2603208d4a3ab4/src/test/run-make-fulldeps/issue-51671)

*Deviations from the RFC:*

- The name of the attribute [was changed] from `panic_implementation` to `panic_handler`. [Proposal](https://github.com/rust-lang/rust/issues/44489#issuecomment-409377261)

[was changed]: https://github.com/rust-lang/rust/issues/44489#issuecomment-415140224

*Remaining known bugs:*

None left. There were a few bugs related to symbol visibility and codegen
options like LTO, but they have been fixed.

**Comment by Mark-Simulacrum** [comment]
Can @rust-lang/lang assign someone or update us on the current state here? Was this intended to be stabilized in RC? If so we need a stabilization PR ASAP.

**Comment by japaric** [comment]
@Mark-Simulacrum 

> Was this intended to be stabilized in RC?

Yes, it would be ideal to have this in RC1 (1.30-beta). If it doesn't make the beta cutoff, could we backport it?

> If so we need a stabilization PR ASAP.

[Stabilization PR](https://github.com/rust-lang/rust/pull/51366)

[Reference PR](https://github.com/rust-lang-nursery/reference/pull/362)

Updating / finishing them up right now.

**Comment by sanmai-NL** [comment]
@Centril: Can you please update the OP with the current status (stabilization etc.)?

**Comment by Centril** [comment]
@sanmai-NL I will defer this to @japaric since they are most up to date as to the status.

**Comment by Centril** [comment]
@sanmai-NL ok, I ticked some boxes which seemed fulfilled; but I don't have the time to do a more in-depth review right now. :)

**Comment by yodaldevoid** [comment]
@Centril @japaric about the the above message as @bors is a bot.

**Comment by curiousleo** [comment]
> @Centril @japaric about the the above message as @bors is a bot.

Lol thanks!

**Comment by eddyb** [comment]
It's not even the bot (@bors) closing the issue, this is a GitHub feature.
When a PR with the right keywords in its description (#51366 in this case) gets merged, issues can be closed automatically by GitHub.

@curiousleo I think the problem there is the unstable feature `panic_info_message` didn't get its own tracking issue, but I'm not sure.

Also a bit surprised that @andreeaflorescu didn't get a reply in almost a year...

**Comment by SimonSapin** [comment]
https://github.com/rust-lang/rust/pull/66771 changes the tracking issue of `PanicInfo::message` to https://github.com/rust-lang/rust/issues/66745, which also has some discussion of what to do about it.

**Comment by curiousleo** [comment]
> #66771 changes the tracking issue of `PanicInfo::message` to #66745, which also has some discussion of what to do about it.

Thanks for following up! That's massively appreciated. I'll keep an eye on those two issues.
